src/dbn_struct.jl

```diff
@@ -0,0 +1,200 @@
+Reference to one or more named nodes in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNOut, m, s, a, rng)` always takes `s` and `a` arguments (the inputs to the entire DBN) while `gen(::DBNVar, m, depargs..., rng)` takes a variable number of arguments (one for each dependency).
+
+`DBNOut` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNOut{names} end
+
+DBNOut(name::Symbol) = DBNOut{name}()
+DBNOut(names...) = DBNOut{names}()
+
+struct DBNDef{N<:NamedTuple, D<:NamedTuple}
+    nodes::N
+    deps::D # values are tuples of DBNVars
+end
+
+node(d::DBNDef, name::Symbol) = d.nodes[name]
+depvars(d::DBNDef, name::Symbol) = d.deps[name]
+depnames(d::DBNDef, n::Symbol) = map(name, depvars(d, n))
+nodenames(d::DBNDef) = keys(d.nodes)
+depstype(DBN::Type{D}) where D <: DBNDef = DBN.parameters[2]
+
+@generated function add_node(d::DBNDef, n::DBNVar{name}, node, deps) where name
+    # deps is a tuple of DBNVars
+    quote 
+        @assert !haskey(d.nodes, name) "DBNDef already has a node named :$name"
+        DBNDef(merge(d.nodes, ($name=node,)), merge(d.deps, ($name=deps,)))
+    end
+end
+
+function add_node(d::DBNDef, n::Symbol, node, deps::NTuple{N,Symbol}) where N
+    return add_node(d, DBNVar(n), node, map(DBNVar, deps))
+end
+
+"""
+Create a list of name=>deps pairs sorted so that dependencies come before dependents.
+"""
+function sorted_deppairs end # this is implemented below
+
+# standard DBNs
+function mdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = map(DBNVar, (:s, :a)),
+            r = map(DBNVar, (:s, :a, :sp)),
+           )
+          )
+end
+
+function pomdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            o = DistributionDBNNode(observation),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = map(DBNVar, (:s, :a)),
+            o = map(DBNVar, (:s, :a, :sp)),
+            r = map(DBNVar, (:s, :a, :sp, :o)),
+           )
+          )
+end
+
+"""
+    DBNStructure(::Type{M}) where M <: Union{MDP, POMDP}
+
+Trait of an MDP/POMDP type for describing the structure of the dynamic Baysian network.
+
+# Example
+
+    struct MyMDP <: MDP{Int, Int} end
+    POMDPs.gen(::MyMDP, s, a, rng) = (sp=s+a+rand(rng, [1,2,3]), r=s^2)
+
+    # make a new node, delta_s, that is deterministically equal to sp - s
+    function POMDPs.DBNStructure(::Type{MyMDP})
+        dbn = mdp_dbn()
+        return add_node(dbn, :delta_s, FunctionDBNNode((m,s,sp)->sp-s), (:s, :sp))
+    end
+
+    gen(DBNOut(:delta_s), MyMDP(), 1, 1, Random.GLOBAL_RNG)
+"""
+function DBNStructure end
+
+DBNStructure(::Type{M}) where M <: MDP = mdp_dbn()
```

Comment:
Do you find it confusing because "Structure" makes you think of `struct`, or for some other reason?

I still think `DBNStructure` is the best name because it does not completely define the DBN - you need the pomdp object as well. @MaximeBouton @Shushman , what do you think?

It is capitalized to follow https://docs.julialang.org/en/v1/manual/methods/#Trait-based-dispatch-1

Reply:
> the fact that it is caps suggests that there is a corresponding type with the same name.

I just renamed `DBNDef` to `DBNStructure`

Reply:
Capitalization: I'm fine with capitalization if that's how it's done in Base. I was just used to `SimpleTraits.jl` where they start with lower case.
Yes, `Structure` makes me think of `struct`. I think I just found it inconsistent because the RHS only says `[po]mdp_dbn` not `[po]mdp_dbnstructure` and that there is never actually a `DBN` type. I understand that the full DBN is only known if you also have the [PO]MDP instance (which gives you the transition likelihoods on the edges) but one could also argue that the "network" is only defined by the layout/topology.

However, this was a very nitpicky comment of mine and I consider this detail that we should probably not waste much energy on.

Reply:
> > the fact that it is caps suggests that there is a corresponding type with the same name.
> 
> I just renamed `DBNDef` to `DBNStructure`

Okay, I'm happy now :tada: